internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher_test.go

```diff
@@ -6,74 +6,1205 @@ package githubupstreamwatcher
+						Phase: v1alpha1.GitHubPhaseReady,
+						Conditions: []metav1.Condition{
+							buildClientCredentialsObtainedTrue(t),
+							buildGitHubConnectionValidTrue(t, *validFilledOutIDP.Spec.GitHubAPI.Host),
+							buildHostValid(t, metav1.ConditionTrue, *validFilledOutIDP.Spec.GitHubAPI.Host),
+							buildOrganizationsPolicyValidTrue(t),
+							buildTLSConfigurationValid(t, metav1.ConditionTrue, *validFilledOutIDP.Spec.GitHubAPI.Host),
+						},
+					},
+				},
+			},
+		},
+		{
+			name:    "happy path with minimal fields",
+			secrets: []runtime.Object{goodSecret},
+			githubIdentityProviders: []runtime.Object{
+				validMinimalIDP,
+			},
+			wantResultingCache: []*upstreamgithub.ProviderConfig{
+				{
+					Name:               "minimal-idp-name",
+					ResourceUID:        "minimal-uid",
+					Host:               fmt.Sprintf("https://%s", goodServerDomain),
+					UsernameAttribute:  "login",
+					GroupNameAttribute: "slug",
+					OAuth2Config: &oauth2.Config{
+						ClientID:     "some-client-id",
+						ClientSecret: "some-client-secret",
+					},
+					OrganizationLoginPolicy: "AllGitHubUsers",
+					AuthorizationURL:        fmt.Sprintf("https://%s/login/oauth/authorize", goodServerDomain),
+					HttpClient:              buildPretendHttpClient(t, goodServer),
+				},
+			},
+			wantResultingUpstreams: []v1alpha1.GitHubIdentityProvider{
+				{
+					ObjectMeta: validMinimalIDP.ObjectMeta,
+					Spec:       validMinimalIDP.Spec,
+					Status: v1alpha1.GitHubIdentityProviderStatus{
+						Phase: v1alpha1.GitHubPhaseReady,
+						Conditions: []metav1.Condition{
+							buildClientCredentialsObtainedTrue(t),
+							buildGitHubConnectionValidTrue(t, *validMinimalIDP.Spec.GitHubAPI.Host),
+							buildHostValid(t, metav1.ConditionTrue, *validMinimalIDP.Spec.GitHubAPI.Host),
+							buildOrganizationsPolicyValidTrue(t),
+							buildTLSConfigurationValid(t, metav1.ConditionTrue, *validMinimalIDP.Spec.GitHubAPI.Host),
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "multiple idps - two good, one invalid",
+			secrets: []runtime.Object{
+				goodSecret,
+				func() runtime.Object {
+					otherSecret := goodSecret.DeepCopy()
+					otherSecret.Name = "other-secret-name"
+					otherSecret.Data["clientID"] = []byte("other-client-id")
+					otherSecret.Data["clientSecret"] = []byte("other-client-secret")
+					return otherSecret
+				}(),
+			},
+			githubIdentityProviders: []runtime.Object{
+				validFilledOutIDP,
+				func() runtime.Object {
+					otherIDP := validFilledOutIDP.DeepCopy()
+					otherIDP.Name = "other-idp-name"
+					otherIDP.Spec.Client.SecretName = "other-secret-name"
+					return otherIDP
+				}(),
+				func() runtime.Object {
+					invalidIDP := validFilledOutIDP.DeepCopy()
+					invalidIDP.Name = "invalid-idp-name"
+					invalidIDP.Spec.Client.SecretName = "no-secret-with-this-name"
+					return invalidIDP
+				}(),
+			},
+			wantResultingCache: []*upstreamgithub.ProviderConfig{
```

Comment:
Nice, this is better than some of our other tests where we just `count` the cache, rather than assess what is in it.